Update get.go to use server-side printing

juanvallejo opened the pull request:
Addresses part of https://github.com/kubernetes/kubernetes/issues/58536
Adds support for server-side changes implemented in https://github.com/kubernetes/kubernetes/pull/40848 and updated in https://github.com/kubernetes/kubernetes/pull/59059

@deads2k per our discussion, opening this as a separate PR.
This wires through a per-request use of `as=Table;...` header parameters 
using the resource builder from the `kube get` command.



#### Items to consider going forward:

- [ ] Figure out how to handle sorting when dealing with multiple Table objects from the server
- [ ] Figure out sorting when dealing with a mixed response from the server consisting of Tables and normal resources (`--sort-by` is handled in this PR by falling back to old behavior)
-  [ ] Filtering: How should we filter Table objects? Separate filter for rows? Filter on jsonpath? We have access to partial object metadata for each table row - not enough to know how to filter pods, for example but we could request that the original object be included along with each Table.Row by adding an `includeObject` param in the client request.

#### Resources that do not yet support Table output

- [x] Namespaces
- [x] Services
- [ ] Service catalog resources: https://github.com/kubernetes-incubator/service-catalog/blob/master/pkg/apis/servicecatalog/v1beta1/types.go

**Release note**:
```release-note
NONE
```

juanvallejo commented on pkg/kubectl/resource/helper.go:
```diff
@@ -72,6 +73,15 @@ func (m *Helper) List(namespace, apiVersion string, export bool, options *metav1
+		table := &metav1alpha1.Table{}
+		err := req.
+			AsTable(metav1.GroupVersion{}).
+			Do().Into(table)
```
@deads2k I'm thinking it'd be best to just have a codec that handles this, however I'm not too familiar with this part of the codebase: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/negotiated_codec.go#L34

deads2k commented on staging/src/k8s.io/client-go/rest/request.go:
```diff
@@ -162,6 +163,26 @@ func (r *Request) Suffix(segments ...string) *Request {
 	return r
 }
 
+// AsTable adds the "as=Table;v=v1alpha1;g=meta.k8s.io" parameter to the Accept header
```
Note that this method is simply a pretty skin for `SetHeader`

deads2k commented on staging/src/k8s.io/client-go/rest/request.go:
```diff
@@ -162,6 +163,26 @@ func (r *Request) Suffix(segments ...string) *Request {
+	tableParam := fmt.Sprintf(";as=Table;v=%s;g=%s", gv.Version, gv.Group)
+
+	if len(r.content.AcceptContentTypes) > 0 {
+		r.SetHeader("Accept", r.content.AcceptContentTypes+tableParam)
```
Ah, it's not.  It seems like we'll eventually want a way to add or set content types, but this could hold for a while.

deads2k commented on pkg/kubectl/resource/helper.go:
```diff
@@ -63,7 +64,7 @@ func (m *Helper) Get(namespace, name string, export bool) (runtime.Object, error
 }
 
-func (m *Helper) List(namespace, apiVersion string, export bool, options *metav1.ListOptions) (runtime.Object, error) {
+func (m *Helper) List(namespace, apiVersion string, export, requestTable bool, options *metav1.ListOptions) (runtime.Object, error) {
```
You'll need this for `Get` too.

juanvallejo replied:
Thanks, wired it through to `Get` as well

smarterclayton replied:
No.  Yuck.  This should be a separate flag on Helper that changes the request type.  We'll have other things we change to, like PartialObjectMetadata.

You also shouldn't have to be parsing into things, you just need to add the table type to the scheme

smarterclayton replied:
See https://github.com/kubernetes/kubernetes/pull/58404/commits/c672a854b7f58e80e6a4f7b1b05d51c695fb2a76#diff-92d4543eda898d1cb24544d3399bebb3R41 and https://github.com/kubernetes/kubernetes/pull/58404/commits/c672a854b7f58e80e6a4f7b1b05d51c695fb2a76#diff-1bc8485222e19c02f9aac7ecbc1cdfd6R50 for adding to the scheme

smarterclayton replied:
Regarding setting this, we should be transforming request like https://github.com/kubernetes/kubernetes/pull/58404/commits/c672a854b7f58e80e6a4f7b1b05d51c695fb2a76#diff-b5780bd984478baab9841efd2c817719R267

And then handling the presence of a Table gracefully in the returned get.go code.

deads2k commented on staging/src/k8s.io/client-go/rest/request.go:
```diff
@@ -162,6 +163,26 @@ func (r *Request) Suffix(segments ...string) *Request {
 }
 
+// AsTable adds the "as=Table;v=v1alpha1;g=meta.k8s.io" parameter to the Accept header
+func (r *Request) AsTable(gv metav1.GroupVersion) *Request {
```
@smarterclayton this looks like the correct spot to set the options, since I wouldn't want an entirely new client for one request.  The rest of the pull seems to flow from here.  In the builder, we could argue things like enum over bool and the like, but given how `get.go` works, we need to set the value from `pkg/kubectl/resource/helper.go`

smarterclayton replied:
I don't think request should know anything about Table.  This should just be SetHeader, only higher levels know about Tables.  PartialObjectMetadata will also come through this path and I don't want `AsPartialObjectMetadata`

fabianofranz commented:
/unassign

soltysh commented:
/assign

juanvallejo commented:
@soltysh this has been rebased.
Intend to use it to implement client-side items from: https://github.com/kubernetes/kubernetes/issues/58536

juanvallejo commented:
@smarterclayton PTAL
This addresses the first two checkboxes under `Client:` from https://github.com/kubernetes/kubernetes/issues/58536

juanvallejo commented:
@smarterclayton thanks for the feedback, comments addressed

juanvallejo commented:
cc @soltysh @mfojtik

juanvallejo commented:
@smarterclayton friendly ping

soltysh commented:
/ok-to-test

soltysh commented on pkg/kubectl/cmd/resource/get.go:
```diff
@@ -284,6 +301,15 @@ func (options *GetOptions) Run(f cmdutil.Factory, cmd *cobra.Command, args []str
 	for ix := range infos {
+		if options.ServerPrint {
+			table, err := options.decodeIntoTable(f.JSONEncoder(), infos[ix].Object)
+			if err == nil {
```
Maybe add a debug statement at level 2 with the error and info that you're falling back to generic. It might save us some debugging cycles.

soltysh commented on pkg/kubectl/cmd/resource/get.go:
```diff
@@ -666,6 +707,10 @@ func addOpenAPIPrintColumnFlags(cmd *cobra.Command) {
 }
 
+func addServerPrintColumnFlags(cmd *cobra.Command) {
+	cmd.Flags().Bool(useServerPrintColumns, true, "If true, have the server return the appropriate table output. Supports extension APIs and CRD. Experimental.")
```
IIRC it should not be turned on by default.

soltysh replied:
In it clearly states 'Have a flag to disable server side rendering that defaults to false' in #58536

soltysh reviewed (approved):
/lgtm

smarterclayton commented:
I'm ok with default on to get soak time in 1.10.  Let's discuss in the sig meeting what the final decision is for 1.10 GA (we can always flip it back to default to false then)

juanvallejo commented:
@smarterclayton @soltysh test-cmd tests are failing on `--sort-by` tests. We could sort the rows on each Table, but not sure how we should handle multiple tables / a mixture of Tables and non-table objects.

soltysh commented:
I've talked with @juanvallejo my proposition for now is to fallback to the old printing when sorting comes into play. This way we'll move this forward, don't break anything and wait for @smarterclayton's pr fixing the biggest issue which is mixing tabled object with old printing mechanism. There are a few issues with sorting that needs better addressing and I don't see we can make it right away.

juanvallejo commented on pkg/kubectl/cmd/resource/get.go:
```diff
@@ -341,7 +376,7 @@ func (options *GetOptions) Run(f cmdutil.Factory, cmd *cobra.Command, args []str
 			}
 
-			printer, err = f.PrinterForMapping(printOpts)
+			printer, err = f.PrinterForOptions(printOpts)
```
@deads2k fyi. Had to change this as `NewVersionedPrinter` inside of PrinterForMapping was returning an error `no kind is registered for the type v1beta1.Table`.

deads2k commented:
/approve

juanvallejo commented:
/retest

deads2k commented:
@juanvallejo failures look consistent.  You'll want to figure out what is broken.

juanvallejo commented:
/test pull-kubernetes-unit

juanvallejo commented:
@deads2k thanks, tests are now passing, wondering it's good to lgtm?

deads2k commented:
> @deads2k thanks, tests are now passing, wondering it's good to lgtm?

I'm ok with it in concept and the idea of a silent fall-back matches the openapi printing we had before, so I'm ok with the default in concept as well, but I haven't reviewed the code in detail.  I figured that @soltysh and @pwittrock would be closer to it.

juanvallejo commented:
I have tested locally to ensure that these changes support a one version skew against an apiserver.
I am able to successfully run `kubectl get` with serverside-printing enabled against 1.9 and 1.10 apiservers.
I am not able to run the same command against 1.8 or 1.7 apiservers (unless I also specify `--experimental-server-print=false`)

Results below:

- 1.9 and 1.10 (success): https://gist.github.com/juanvallejo/9577cfa079a45f31ce2e0eccdb6793ad
  - Able to retrieve resources that support tabular responses
  - Able to fallback to a non-tabular response for resources that do not yet support tabular responses

- 1.8 (error): https://gist.github.com/juanvallejo/283954e9f83ed3e6f0398bbec3ca5502
  - Unable to retrieve resources that support tabular responses
  - Unable to retrieve resources that do not yet support tabular responses

- 1.7 (error): https://gist.github.com/juanvallejo/e9662ba4cb2fa926526e8d00b6cc48b3
  - Unable to retrieve resources that support tabular responses
  - Unable to retrieve resources that do not yet support tabular responses

soltysh commented:
@juanvallejo there are two possible solutions to this being compatible with 1.8 and earlier versions. We either check the server version and invoke the `TransformRequests` or we do 2 requests with transformation and if that fails another without it.  /cc @pwittrock 
Personally I'm more in favor of the first approach.

soltysh commented:
With a simple check like this:
```go
```

smarterclayton commented:
Please don't do version checking like that.

If this is something we can backport to 1.8 (it's probably the fall through version checking) that's better.

soltysh commented:
@smarterclayton we've discussed this in length during scrum, the 2 requests I've proposed is the way to go.

juanvallejo committed:
handle Table response in client

This patch adds support for the "server-side GET operation"
introduced by pull/40848 and proposed by kubernetes/community#363.

soltysh reviewed (approved):
/lgtm

smarterclayton commented:
Sorting and printing should be using the table but with full objects returned via the print API (`?includeObject=Object`) and then doing the sort on nested, as should custom columns.

soltysh commented:
@smarterclayton I've added a point about that to the general server side printing issue.
